### DISCUSSION TITLE
Configured and self-hosted outline. Getting a 502 Bad Gateway Error at the end of the setup

### DISCUSSION THREAD

**Discussion by AshwinPrasadRao** [outline/outline#6779] [Self Hosting]
Hello everyone, 

I set up Outline on DigitalOcean for my organisation. All the docker containers are running and there are no errors in docker compose logs. However, I see some errors in the Nginx logs:

```
recv() failed (104: Connection reset by peer) while reading response header from upstream, client: {my ip}, server: {my domain}, request: "GET /favicon.ico HTTP/1.1", upstream: "http://127.0.0.1:3000/favicon.ico", host: {my droplet ip}

upstream prematurely closed connection while reading response header from upstream, client: {my ip}, server: {my domain}, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: {my droplet ip}
```

I have tried setting user: root in the docker-compose.yml file and it didn't solve the issue. Would really appreciate any help here.

**Comment by przemokam** [comment]
Try setting: 
proxy_pass http://localhost:3000; 
in your /etc/nginx/conf.d/outline.conf file. It helped me with 502 error.

**Reply by oflarcade** [reply]
unfortunately this didn't help

**Comment by oflarcade** [comment]
Hey @AshwinPras 
I am encountering the same issue, did you manage to find a fix for this issue ?

**Reply by AshwinPrasadRao** [reply]
Hello @oflarcade, what eventually worked for me was running `sudo certbot` and choosing `Attempt to reinstall this existing certificate`.

**Comment by ghost** [comment]
TLDR: The main goal is to ensure that Outline is listening on port 3000. This was the ultimate challenge. Once you confirm it is loading over port 3000, you'll start to see other minor issues that need to be addressed. This is just my experience, though.

---

I recently resolved this issue as I couldn't find much information online. To note, I am self-hosting this on Ubuntu 24.02 with the latest versions of the dependencies mentioned in their [From Source](https://docs.getoutline.com/s/hosting/doc/from-source-BlBxrNzMIP) documentation and Nginx installed. This may not resolve the issue for you, but it's a summary of what I did that led to it working and not necessarily in this order:

---

1. **Yarn Permissions:**
   - Initially, the `yarn` command failed to execute due to global permission issues. To resolve this, we ensured that Yarn was installed correctly and accessible by the Outline service user.
   - Solution: I reinstalled Yarn with the correct permissions using `npm install --global yarn` and ensured that the Outline service user had the necessary permissions to execute it.

2. **Database User Authentication:**
   - The 502 error was also linked to issues with the PostgreSQL database user I created. The authentication failed due to incorrect or insufficient permissions in the `pg_hba.conf` file.
   - We modified the `pg_hba.conf` file to allow local and remote connections for the database user. This involved setting the correct authentication method, in my case, `md5`, and restarting the PostgreSQL service.
   - Ultimately, we had to set the database user to Postgres itself, perform the migration, and then place it back on the user I created.

3. **Environment Configuration:**
   - We verified and corrected the environment variables, mainly focusing on the `DATABASE_URL` and other related database settings.
   - Additionally, we ensured that the `FORCE_HTTPS` environment variable was turned off to prevent unnecessary redirects that could cause further issues. _This works if you have a similar setup of using an internal proxy that handles your SSL and HTTP to HTTPS redirects. Since all URLs had to be `http`, enabling Web Socket Support is necessary. This also allows the editor on the Documents to load without reloading after typing in the title. This is more for those using the [Nginx Proxy Manager](https://github.com/NginxProxyManager/nginx-proxy-manager?utm_source=nginx-proxy-manager) or similar for example. You will see Web Socket errors in your browser's console._

4. **Database Migration Kept Failing:**
   - The database migration initially failed due to permission issues. After resolving the authentication problems from Section 2 above, we were then able to run the database migration.
     ```bash
     yarn sequelize db:migrate
     ```

5. **Outline Service Configuration:**
   - The Outline application was configured as a system service. We corrected the `User` directive in the service configuration to ensure it ran under the correct user with appropriate permissions.
   - The service was restarted, and the application ran without further issues.

6. **Redirect Loop Fix:**
   - After resolving the primary issues, a redirect loop was observed, which was fixed by ensuring the `FORCE_HTTPS` environment variable was false.

7. **Other Things:**
I created an alias for the systems log of the outline service I created. That way, it was easy to view additional errors. If you are not setting it up as a running service, then installing PM2 can help. Nginx logs only showed one error throughout most of the process, which caused some delays in resolving it. I tried this method initially before switching to making it a service. I found the system logs more informative than PM2.

**Reply by AshwinPrasadRao** [reply]
Thanks for the detailed steps, @aCyberVoid. To add to your list, running `sudo certbot` and `Attempt to reinstall this existing certificate` solved the issue for me.
